### PR TITLE
Changes based on comments from Gunnar H: cps spelling nit

### PR DESCRIPTION
**Pull request by cdh4u**

### CODE REVIEW

**Review by gunnarhm — APPROVED** [review]

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 1]
```diff
@@ -135,42 +131,45 @@
...
-          'cps' attribute parameter with a attribute parameter value of 20:
-        </t>
-        <t>
-          a=dcsa:1 fmtp:- cps=20
+        <section anchor="sec.sdp.dcsa.trans" title="Maximum Character Transmission">
+          <t>
+            A 'dcsa' attribute can contain the SDP fmtp attribute used to indicate
+            a maximum character transmission rate <xref target="RFC4103"/>. The 'cps'
+            attribute parameter is used indicate the maximum character transmission rate
+            that the endpoint that includes the attribute is able to handle. The 'format' attribute
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: Should it be "receive" instead of "handle"? Because it is reception that can be limited.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 2]
```diff
@@ -135,42 +131,45 @@
...
+            offers and answers.
+          </t>
+          <t>
+            NOTE: The 'cps' attribute parameter is especially useful when a T.140 data channel
+            endpoint is acting as a gateway [<xref target="sec.gw"/>] and is interworking with
+            a T.140 transport mechanism that have restrictions on how many characters can be sent
+            per second.
+          </t>
+        </section>
+        <section anchor="sec.sdp.dcsa.lan" title="Real-time Text Conversation Languates">
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: "Languates" in the title should be "Languages"

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 3]
```diff
@@ -135,42 +131,45 @@
...
+         <t>
+            A 'dcsa' attribute can contain the SDP hlang-send and hlang-recv attributes
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: A dcsa ... sounds as if one attribute can contain both hlang-send and hlang-recv attributes simultaneously. But one dcsa is needed for each. maybe start "The 'dcsa'attributes can contain...."

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 4]
```diff
@@ -185,6 +184,41 @@
...
+    a=sctp-port 5000
+    a=dcmap:1 label="ACME customer service";subprotocol="t140"
+    a=dcsa:1 fmtp:-
+    a=dsca:1 hlang-send:es eo
+    a=dsca:1 hlang-recv:es eo
+
+Answer:
+
+    m=application 911 UDP/DTLS/SCTP webrtc-datachannel
+    c=IN IP6 2001:db8::3
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: I think the answer should have another c= address than the offer. please invent another address.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 5]
```diff
@@ -239,6 +273,46 @@
...
+        </section>
+        <section anchor="sec.t140.loss" title="Loss of T140blocks">
+          <t>
+            In case of network failure or congestion, T140 data channels might get lost.
+            If this happens but the session sustains, it is RECOMMENDED that a low number of 
+            retries are made to reestablish the T140 data channels. If reestablishment of
+            the T140 data channel is successful, an implementation MUST evaluate if any 
+            T140blocks were lost. Retransmission of already transmitted T140blocks MUST be
+            avoided, and missing text markers <xref target="T140ad1"/> SHOULD be inserted in the received 
+            data stream where loss is detected or suspected and presented to the user.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: The sentence got too complex. I suggest to delete ""and presented to the user"

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 6]
```diff
@@ -239,6 +273,46 @@
...
+          <t>
+            An implementation needs to take the user requirements for smooth flow and low
+            latency in real-time text conversation into consideration when assigning a buffer time.
+            It is RECOMMENDED to use the default transmission interval of 300 milliseconds
+            <xref target="RFC4103"/>, or lower, also for T.140 data channels.
+          </t>
+        </section>
+        <section anchor="sec.t140.loss" title="Loss of T140blocks">
+          <t>
+            In case of network failure or congestion, T140 data channels might get lost.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: Is 'get lost' the best term?  I have tried to find a good specification for it. looked in sdpneg, in data-channel and in W3C API.  "torn down with error indication" may be a suitable term. What do you think?

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 7]
```diff
@@ -239,6 +273,46 @@
...
+          <t>
+            An implementation needs to take the user requirements for smooth flow and low
+            latency in real-time text conversation into consideration when assigning a buffer time.
+            It is RECOMMENDED to use the default transmission interval of 300 milliseconds
+            <xref target="RFC4103"/>, or lower, also for T.140 data channels.
+          </t>
+        </section>
+        <section anchor="sec.t140.mul" title="Multi-party Considerations">
+          <t>
+            If an implementation needs to support multi-party scenarios, implementations need
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: Singular - plural conflict. It starts "if an implementation.... and continues "implementations need".  I suggest to start the second phrase ", the implementation needs"

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 8]
```diff
@@ -239,6 +273,46 @@
...
+            to support multiple simultaneous T.140 data channels, one for each remote party. At
+            the time of writing this document, this is true even in scenarios where each participants
+            communicate via a centralized conference server. The reason is that, unlike RTP media,
+            WebRTC data channels and the T.140 protocol do not support the indication of the source
+            of T.140 data. The SDP dcmap attribute label attribute parameter (<xref target="sec.sdp.dcmap"/>
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: do we need to say "SDP dcmap attribute label attribute parameter"? It gets complex.
Can it do with "dcmap label attribute" ?

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 9]
```diff
@@ -239,6 +273,46 @@
...
+            can be used to provide additional information about each T.140 data channel, and help
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: We learned that it is only the offering party who can set the value, while the answer must apparently repeat the same value.  (according to sdpneg examples)
 I suggest indicating that by changing "additional information" to "additional information from the offering party"

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 10]
```diff
@@ -239,6 +273,46 @@
...
+            implementations to distinguish between them.
+          </t>
+          <t>
+            NOTE 3: Future extensions to T.140, or to the T140block, might allow indicating the source
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: I don't think this NOTE need to be numbered. I expect it is needed only if we have more than one note in a section, but that needs to be checked with an IETF authoring guide.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 11]
```diff
@@ -104,24 +104,20 @@
...
-            <t>"subprotocol=" "t140"</t>
-          </list>
+          The offerer and answerer MUST include the subprotocol attribute parameter,
+          with a "t140" parameter value, in the dcmap attribute value.
         </t>
         <t>
-          The offerer and answerer MUST NOT include the max-retr, max-time and ordered
-          attribute parameters in the dcmap attribute.
+          The offerer and answerer MAY, based on local policy, include the priority
+          attribute parameter and the label attribute parameter in the dcmap attribute value.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: Is "label" really included based on local policy? I think it is when needed by the protocol.  Can we delete the "based on local policy" for both, or do you need to reorder the wording?

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 12]
```diff
@@ -264,7 +338,19 @@
...
+        <t>
+          This document updates RFC 8373, by defining how the SDP hlang-send and hlang-recv attributes are used for
+          the "application/webrtc-datachannel" media type.
+        </t>
+        <t>
+          SDP offerers and answerers MUST NOT include the attributes directly in the m= section associated with the
+          'application/webrtc-datachannel' media type. Instead, the attributes MUST be associated with
+          individual data channels, using the SDP 'dcsa' attribute. A specification that defines a subprotocol
+          that uses the attributes MUST specify the modality for that subprotocol, or how to retreive the 
+          modality if the subprotocol supports multiple modalifies.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: "modalifies" sb "modalities"

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 13]
```diff
@@ -304,7 +304,7 @@ Answer:
             the time of writing this document, this is true even in scenarios where each participants
             communicate via a centralized conference server. The reason is that, unlike RTP media,
             WebRTC data channels and the T.140 protocol do not support the indication of the source
-            of T.140 data. The SDP dcmap attribute label attribute parameter (<xref target="sec.sdp.dcmap"/>
+            of T.140 data. The SDP dcmap attribute label attribute parameter (<xref target="sec.sdp.dcmap"/>)
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: SDP dcmap attribute label attribute parameter   is a complex expression, hard to follow. 
can we call it "dcmap label attribute"?

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 14]
```diff
@@ -318,24 +318,51 @@ Answer:
...
+          <t>
+            For each T.140 data channel there is an RTP stream for real-time text <xref target="RFC4103"/> . 
+            Redundancy is by default declared and used on RTP stream. On the T.140 data channel
+            there is no redundancy, but the reliable property <xref target="I-D.ietf-mmusic-data-channel-sdpneg"/>
+            of T.140 the data channel is set.
+          </t>
+          <t>
+            During a normal text flow, T140blocks received from one network are forwarded towards the other network.
+            Keep-alive traffic is implicit on the T140 data channel. <xref target="RFC4103"/> does not define
+            keep-alive traffic for the RTP stream.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: Keep-alive traffic is usually needed on the RTP side, even if RFC 4103 does not mention it. 
Proposal: replace the sentence about does not define.   "Keep-alive traffic MAY need to be extracted from incoming RTP traffic and generated on outgoing RTP traffic according to procedures used on the RTP side."

File: draft-holmberg-mmusic-t140-usage-data-channel.xml [hunk 15]
```diff
@@ -318,24 +318,51 @@ Answer:
...
+            It is RECOMMENDED that the gateway uses the same transmission interval on both the T140 data channel and the RTP stream,
+            if possible. That will reduce the delay caused by buffering.
+          </t>
+          <t>
+            If the gateway detects or suspects loss of data on the RTP stream, the gateway gateway SHOULD insert the T.140 missing
+            text marker <xref target="T140ad1"/> in the data sent on the T.140 data channel. 
+          </t>
+          <t>
+            If the gateway detects that the T.140 data channel has failed, once the data channel has been reestablished
+            the gateway SHOULD insert the T.140 missing text marker <xref target="T140ad1"/> in the data sent on the RTP stream.
```
Comment (gunnarhm) [inline draft-holmberg-mmusic-t140-usage-data-channel.xml]: This is fuzzy. I think directions need to be included in the sentences. And consideration if anything was lost.

**Review by gunnarhm — CHANGES_REQUESTED** [review]
I have reviewed all change proposals and commented where I propose changes.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 1 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: I will change to "receive".

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 2 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 3 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 4 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: I was going to do that, but forgot :) Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 5 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 6 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: I will look into it. I agree that "get lost" is not a good term.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 7 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 8 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: "Attribute" refers to an SDP attribute, and label is not an attribute. We have used this type of wording also in other documents, and I think we should use the complete name at least on first occurrence. I will see if it can be simplified elsewhere.

(Note that there is also an SDP label attribute, defined in RFC 4574, but that is not what we are talking about in the draft, and that needs to be clear in the text.)

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 9 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix. I could also add a note, referring to the RTCWEB data channel text saying that the value must be the same in both directions.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 10 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: I have seen all kind of variations when reviewing drafts, so I am not sure whether there are any rules. If there are, I am sure the RFC Editor will fix it.

But, I can remove the number for now.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 11 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: I can remove "based on local policy".

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 12 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 13 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: See my earlier comment on this.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 14 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Will fix.

File: draft-holmberg-mmusic-t140-usage-data-channel.xml (hunk 15 above)
Comment (cdh4u) [reply draft-holmberg-mmusic-t140-usage-data-channel.xml]: Doesn't "insert the T.140 missing text marker in the data sent on the RTP stream" indicate the direction? But, to make it more clear, I can add ", towards the RTP endpoint."
